File: src-ml/lin.ml
module Lin = struct
    let swap a i j =
        let tmp = a.(i) in
        a.(i)<-a.(j);
        a.(j)<-tmp;;

    let arrsub a b fac ibegin iend=
        for i=ibegin to iend-1 do
            a.(i)<-a.(i) -. fac *. b.(i)
        done;;

    let prod a b ibegin =
        let r = ref 0.0 in
        for i = ibegin to (Array.length a)-1 do
            r := !r +. a.(i) *. b.(i)
        done;
        !r;;

    let scale_row factor ai jbegin jend b i =
        for j = jbegin to jend-1 do
            ai.(j) <- factor *. ai.(j)
        done;
        b.(i) <- factor *. b.(i);;

    let scale_array a b=
        let absmax ai =
            Array.fold_left (fun x aij->max x (abs_float aij)) 0.0 ai in
        for i=0 to (Array.length a)-1 do
            let factor = 1.0 /. (absmax a.(i)) in
            scale_row factor a.(i) 0 (Array.length a) b i;
        done;;

    let maxji a i =
        let ajmax=ref (abs_float a.(i).(i)) in
        let maxi=ref i in
        for j=i+1 to (Array.length a)-1 do
            let aji = abs_float a.(j).(i) in
            if aji > !ajmax then (
                ajmax := aji;
                maxi := j
            )
        done;
        !maxi;;

    let pivot a b i=
        let maxi =maxji a i in
        if i<>maxi then (
            swap a i maxi;
            swap b i maxi
            );;

    let update_lower_col a b i blockend=
        let n=(Array.length a) in
        for i1=i to blockend-1 do
            pivot a b i1;
            let ai=a.(i1) in
            scale_row (1.0 /. ai.(i1)) ai (i1+1) blockend b i1;
            for j=i1+1 to n-1 do
                let aj=a.(j) in
                let factor=aj.(i1) in
                arrsub aj a.(i1) factor (i1+1) blockend;
                b.(j) <- b.(j) -. factor *. b.(i1)
            done
        done;;

    let update_upper_row a i blockend=
        let n=(Array.length a) in
        for i1=i to blockend-1 do
            let ai=a.(i1) in
            let factor=1.0 /. ai.(i1)in
            for j=blockend to n-1 do
                ai.(j) <- ai.(j) *. factor
            done;
            for j=i1+1 to blockend-1 do
                let aj=a.(j) in
                arrsub aj ai aj.(i1) blockend n
            done
        done;;

    let update_a a i blockend=
        let n=(Array.length a) in
        let blocksize=4 in
        for j=blockend to n-1 do
            let aj=a.(j) in
            let mend=(blockend-i)/blocksize in
            for m=0 to mend-1 do
                let l=i+m*blocksize in
                let al0=a.(l+0) in
                let al1=a.(l+1) in
                let al2=a.(l+2) in
                let al3=a.(l+3) in
                let ajl0=aj.(l+0) in
                let ajl1=aj.(l+1) in
                let ajl2=aj.(l+2) in
                let ajl3=aj.(l+3) in
                for k=blockend to n-1 do
                    aj.(k)<- aj.(k) -. (ajl0 *. al0.(k) +. ajl1 *. al1.(k)
                        +. ajl2 *. al2.(k) +. ajl3 *. al3.(k))
                done
            done;
            for l=i+blocksize*mend to blockend-1 do
                let al=a.(l) in
                let ajl=aj.(l) in
                for k=blockend to n-1 do
                    aj.(k)<- aj.(k) -. ajl *. al.(k)
                done
            done
        done;;

    let forward_elimination a b=
        let n=Array.length a in
        let blocksize=32 in
        let rec forward_rec i=
            if i<n then
                begin
                    let blockend=min (i+blocksize) n in
                    update_lower_col a b i blockend;
                    update_upper_row a i blockend;
                    update_a a i blockend;
                    forward_rec (i+blocksize)
                end
            in
            forward_rec 0;;

    let back_substitution a b=
        for i = (Array.length b)-1 downto 0 do
            b.(i) <- b.(i) -. (prod a.(i) b (i+1))
        done;
        b;;

    let solve a b=
        scale_array a b;
        forward_elimination a b;
        back_substitution a b;;

end
module type Lin=
    sig
        val solve : 'float array array -> 'float array -> 'float array
end

File: src-java/lin.java
import java.lang.Math.*;

class Lin{
    static void swap(double[][] a, int i, int j)
    {
	double[] tmp=a[i];
	a[i]=a[j];
	a[j]=tmp;
    }
    static void swap(double[] a, int i, int j)
    {
	double tmp=a[i];
	a[i]=a[j];
	a[j]=tmp;
    }
    static void scale_vector(double[] a, int from, int end, double factor)
    {
	for(int i=from; i<end; i++){
	    a[i]*=factor;
	}
    }
    static void scale_matrix_row(double[][] a, double[] b)
    {
	int n=a.length;
	for(int i=0; i<n; i++){
	    double aimax=Math.abs(a[i][0]);
	    for(int j=1; j<n; j++){
		double aij=Math.abs(a[i][j]);
		if(aij>aimax){
		    aimax=aij;
		}
	    }
	    double factor=1.0/aimax;
	    scale_vector(a[i],0,n,factor);
	    b[i]*=factor;
	}
    }
    static void pivot(double[][] a, double[] b, int i)
    {
	int n=a.length;
	double ajimax=Math.abs(a[i][i]);
	int maxj=i;

	for(int j=i+1; j<n; j++){
	    double aji=Math.abs(a[j][i]);
	    if(aji>ajimax){
		ajimax=aji;
		maxj=j;
	    }
	}
	if(maxj!=i){
	    swap(a,i,maxj);
	    swap(b,i,maxj);
	}
    }
    static void update_lower_col(double[][] a,double[] b,int i,int blockend)
    {
	int n=a.length;
	for(int i1=i; i1<blockend; i1++){
	    pivot(a,b,i1);
	    double factor=1.0/a[i1][i1];
	    scale_vector(a[i1],i1+1,blockend,factor);
	    b[i1]*=factor;
	    for(int j=i1+1; j<n; j++){
		factor=a[j][i1];
		for(int k=i1+1; k<blockend; k++){
		    a[j][k]-=factor*a[i1][k];
		}
		b[j]-=factor*b[i1];
	    }
	}
    }
    static void update_upper_row(double[][] a,int i,int blockend)
    {
	int n=a.length;
	for(int i1=i; i1<blockend; i1++){
	    scale_vector(a[i1],blockend,n,1.0/a[i1][i1]);
	    for(int j=i1+1; j<blockend; j++){
		for(int k=blockend; k<n; k++){
		    a[j][k]-=a[j][i1]*a[i1][k];
		}
	    }
	}
    }
    static void forward_elimination(double[][] a,int i,int blockend)
    {
	int n=a.length;
	int blocksize=4;
	for(int j=blockend; j<n; j++){
	    int l;
	    for(l=i; l+blocksize<=blockend; l+=blocksize){
		double ajl0=a[j][l+0];
		double ajl1=a[j][l+1];
		double ajl2=a[j][l+2];
		double ajl3=a[j][l+3];
		for(int k=blockend; k<n; k++){
		    a[j][k]-=ajl0*a[l+0][k]
			+ajl1*a[l+1][k]
			+ajl2*a[l+2][k]
			+ajl3*a[l+3][k];
		}
	    }
	    for(; l<blockend; l++){
		double ajl=a[j][l];
		for(int k=blockend; k<n; k++){
		    a[j][k]-=ajl*a[l][k];
		}
	    }
	}
    }
    static void back_substitution(double[][] a, double[] b)
    {
	int n=a.length;
	for(int i=n-1; i>=0; i--){
	    for(int j=i+1; j<n; j++){
		b[i]-=a[i][j]*b[j];
	    }
	}
    }
    public static void solve(double[][] a, double[] b)
    {
	int n=a.length;
	scale_matrix_row(a,b);
	int blocksize=16;
	for(int i=0; i<n; i+=blocksize){
	    int blockend=Math.min(i+blocksize,n);
	    update_lower_col(a,b,i,blockend);
	    update_upper_row(a,i,blockend);
	    forward_elimination(a,i,blockend);
	}
	back_substitution(a,b);
    }
}
